packages/dara-core/js/shared/global-state-store.tsx

```diff
@@ -0,0 +1,134 @@
+class GlobalStore<TState extends Record<string, any>> {
+    #state: TState;
+
+    #locks: Record<keyof TState, Promise<TState[keyof TState]>>;
+
+    #subscribers: Record<keyof TState, ((val: TState[keyof TState]) => void)[]>;
+
+    constructor() {
+        this.#state = {} as TState;
+        this.#locks = {} as Record<keyof TState, Promise<any>>;
+        this.#subscribers = {} as Record<keyof TState, ((val: TState[keyof TState]) => void)[]>;
+    }
+
+    /**
+     * Clear the store
+     * Removes all values and subscribers
+     */
+    public clear(): void {
+        this.#state = {} as TState;
+        this.#locks = {} as Record<keyof TState, Promise<any>>;
+        this.#subscribers = {} as Record<keyof TState, ((val: TState[keyof TState]) => void)[]>;
+    }
+
+    /**
+     * Get the value for a key.
+     * If the value is being replaced, the promise for the new value is returned.
+     *
+     * @param key - key to get value for
+     */
+    public async getValue<TKey extends keyof TState>(key: TKey): Promise<TState[TKey]> {
+        if (this.#locks[key]) {
+            return this.#locks[key] as Promise<TState[TKey]>;
+        }
+
+        return this.#state[key];
+    }
+
+    /**
+     * Get the value for a key synchronously.
+     * Even if the value is being replaced, the current value is returned.
+     *
+     * @param key - key to get value for
+     */
+    public getValueSync<TKey extends keyof TState>(key: TKey): TState[TKey] {
+        return this.#state[key];
+    }
+
+    /**
+     * Set the value for a key.
+     *
+     * @param key - key to set value for
+     * @param value - value to set
+     */
+    public setValue(key: keyof TState, value: TState[keyof TState]): void {
+        this.#notify(key, value);
+        this.#state[key] = value;
+    }
+
+    /**
+     * Replace the value for a key.
+     * If the value is being replaced, the promise for the new value is returned.
+     *
+     * @param key - key to replace value for
+     * @param fn - function to get the new value
+     */
+    public async replaceValue<TKey extends keyof TState>(
+        key: TKey,
+        fn: () => Promise<TState[TKey]>
+    ): Promise<TState[TKey]> {
+        if (this.#locks[key]) {
+            return this.#locks[key] as Promise<TState[TKey]>;
+        }
+
+        let unlock: (res: TState[TKey]) => void;
+        let unlockError: (err: Error) => void;
+        const lockPromise = new Promise<TState[TKey]>((resolve, reject) => {
+            unlock = resolve;
+            unlockError = reject;
+        });
+
+        this.#locks[key] = lockPromise;
+
+        let result: TState[TKey];
+
+        try {
+            result = await fn();
+            this.#state[key] = result;
+
+            // success - notify subscribers, unlock pending promises and set the value
+            this.#notify(key, result);
+            unlock(result);
+            this.setValue(key, result);
```

Comment:
Isn't this `setValue` doing the same thing as 3 lines above?

Reply:
Good catch, yeah we'll notify twice here. I'll remove the explicit one 👍